src/os/bluestore/BlueStore.h

```diff
@@ -1297,10 +1297,29 @@ class BlueStore : public ObjectStore,
 
   struct OnodeSpace {
   private:
+    struct ghobject_hash_helper_t : public ghobject_t {
+      uint32_t obj_hash;
+
+      ghobject_hash_helper_t(const ghobject_t& o)
+        : ghobject_t(o) {
+        static std::hash<object_t> H;
+        static std::hash<ghobject_t> I;
+        this->obj_hash = H(o.hobj.oid) ^ I(o);
```

Comment:
can we just do o.hobj.hash ^ o.hobj.snap ?  I think that will be sufficient?  Would prefer to avoid rehashing the string if we can!

Reply:
(also, if we do that, then we can just change teh generic hash<> method in hobject.h and avoid this class entirely)

Reply:
That's what `std::hash<ghobject_t>` already [does](https://github.com/rzarzynski/ceph/blob/e227119f502266bff02e54f305da62cbfb655788/src/common/hobject.h#L477). Unfortunately, it's not enough when instances of `hobject_t` are constructed with `CEPH_NOSNAP` like in the case of FIO's `ObjectStore` plugin or `make_pgmeta()`.

We could detect such situations and hash the `o.hobj.oid` only then. Though, I still feel the most pain comes not from the hash computations per se but from memory fetches they depend on. Hashing in the constructor of `ghobject_hash_helper_t` is intended to access them once and as close as possible to `Transaction::iterator::decode_op()`. Additionally, there is [an effort](https://github.com/rzarzynski/ceph/tree/wip-common-sstring-for-ghobject) to replace `std::string` with Ceph's `sstring` to reduce the number of fetches.

`ghobject_hash_helper_t` wrapper has been put to perform the additional computations only in places that actually can benefit from having better differentiated hashes (`std::unordered_map`!).

Reply:
For fio, can we just fix it to use a range of hash values?

And for make_pgmeta(), the hash value should match the pg_t::m_seed, right?

Reply:
Sure, I will fix on the `fio` layer if there isn't any other, prominent usage of `CEPH_NOSNAP`.